pkg/controllers/v1alpha2/mpi_job_controller.go

```diff
@@ -0,0 +1,1158 @@
+		MountPath: configMountPath,
+	})
+	podSpec.Spec.Containers[0] = container
+
+	scriptMode := int32(0555)
+	podSpec.Spec.Volumes = append(podSpec.Spec.Volumes, corev1.Volume{
+		Name: configVolumeName,
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: mpiJob.Name + configSuffix,
+				},
+				Items: []corev1.KeyToPath{
+					{
+						Key:  kubexecScriptName,
+						Path: kubexecScriptName,
+						Mode: &scriptMode,
+					},
+				},
+			},
+		},
+	})
+
+	// set default BackoffLimit
+	if mpiJob.Spec.BackoffLimit == nil {
+		mpiJob.Spec.BackoffLimit = new(int32)
+		*mpiJob.Spec.BackoffLimit = 6
+	}
+
+	return &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      mpiJob.Name + workerSuffix,
+			Namespace: mpiJob.Namespace,
+			Labels:    labels,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(mpiJob, kubeflow.SchemeGroupVersionKind),
+			},
+		},
+		Spec: appsv1.StatefulSetSpec{
+			PodManagementPolicy: appsv1.ParallelPodManagement,
+			Replicas:            &desiredReplicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			ServiceName: mpiJob.Name + workerSuffix,
+			Template:    *podSpec,
+		},
+	}
+}
+
+// newLauncher creates a new launcher Job for an MPIJob resource. It also sets
+// the appropriate OwnerReferences on the resource so handleObject can discover
+// the MPIJob resource that 'owns' it.
+func newLauncher(mpiJob *kubeflow.MPIJob, kubectlDeliveryImage string) *batchv1.Job {
+	launcherName := mpiJob.Name + launcherSuffix
+	labels := map[string]string{
+		labelGroupName:   "kubeflow.org",
+		labelMPIJobName:  mpiJob.Name,
+		labelMPIRoleType: launcher,
+	}
+
+	podSpec := mpiJob.Spec.MPIReplicaSpecs[kubeflow.MPIReplicaTypeLauncher].Template.DeepCopy()
+	// copy the labels and annotations to pod from PodTemplate
+	if len(podSpec.Labels) == 0 {
+		podSpec.Labels = make(map[string]string)
+	}
+	for key, value := range labels {
+		podSpec.Labels[key] = value
+	}
+
+	podSpec.Spec.ServiceAccountName = launcherName
+	podSpec.Spec.InitContainers = append(podSpec.Spec.InitContainers, corev1.Container{
+		Name:  kubectlDeliveryName,
+		Image: kubectlDeliveryImage,
+		Env: []corev1.EnvVar{
+			{
+				Name:  kubectlTargetDirEnv,
+				Value: kubectlMountPath,
+			},
+		},
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      kubectlVolumeName,
+				MountPath: kubectlMountPath,
+			},
+		},
+	})
+	container := podSpec.Spec.Containers[0]
+	container.Env = append(container.Env,
+		corev1.EnvVar{
+			Name:  "OMPI_MCA_plm_rsh_agent",
+			Value: fmt.Sprintf("%s/%s", configMountPath, kubexecScriptName),
+		},
+		corev1.EnvVar{
+			Name:  "OMPI_MCA_orte_default_hostfile",
+			Value: fmt.Sprintf("%s/%s", configMountPath, hostfileName),
+		})
+	// Not assign any resource limits and requests to launcher pod
+	container.Resources.Limits = nil
+	container.Resources.Requests = nil
```

Comment:
the resource maybe can read from job launcher spec, user can specify resource to launcher pod if needed

Reply:
Yes, this was work-in-progress so I haven't pushed all the changes yet. Just updated the PR.